plenoptic/simulate/models/frontend.py

```diff
@@ -47,63 +41,95 @@ class Front_End(nn.Module):
+        self.contrast_scalars = nn.Parameter(torch.rand((1, 2, 1, 1)))
 
-        self.linear = transforms.Compose([pad, linear])
-        self.luminance = transforms.Compose([pad, luminance])
-        self.contrast = transforms.Compose([pad, contrast])
+        # pad all transforms for convolution
+        pad = nn.ReflectionPad2d(self.linear.weight.shape[-1]//2)
+        self.linear_pad = transforms.Compose([pad, self.linear])
+        self.luminance_pad = transforms.Compose([pad, self.luminance])
+        self.contrast_pad = transforms.Compose([pad, self.contrast])
         self.softplus = nn.Softplus()
 
         self.disk_mask = disk_mask
+        self._disk = None  # cached disk to apply to image
 
-    def luminance_normalization(self, x):
+        if pretrained:
+            self._load_pretrained()
 
-        s = self.luminance_scals
-        return torch.div(self.linear(x), (1 + s * self.luminance(x)))
+        if not requires_grad:  # turn off gradient
+            [p.requires_grad_(False) for p in self.parameters()]
 
-    def contrast_normalization(self, x):
+    def _load_pretrained(self):
+        """Load FrontEnd model weights used from Berardino et al (2017)"""
+        state_dict = torch.load(os.path.join(os.path.dirname(__file__), 'weights/FrontEnd.pt'))
+        self.load_state_dict(state_dict)
 
-        s = self.contrast_scals
-        return torch.div(x, 1 + s * torch.sqrt(1e-10 + self.contrast(x ** 2)))
+    def _luminance_normalization(self, x):
+        s = self.luminance_scalars
+        return torch.div(self.linear_pad(x), (1 + s * self.luminance_pad(x)))
 
-    def forward(self, x):
+    def _contrast_normalization(self, x):
+        s = self.contrast_scalars
+        return torch.div(x, 1 + s * torch.sqrt(1e-10 + self.contrast_pad(x**2)))
 
-        image_size = x.shape[-1]
-        x = self.luminance_normalization(x)
-        x = self.contrast_normalization(x)
+    def forward(self, x):
+        x = self._luminance_normalization(x)
+        x = self._contrast_normalization(x)
         x = self.softplus(x)
 
-        if self.disk_mask:
-            self.disk = make_disk(image_size)
+        if self._disk is not None and self._disk.shape == x.shape[-2:]:  # uses cached disk_mask if size matches
+            x = self._disk * x
+
+        elif ((self._disk is not None and self._disk.shape != x.shape[-2:])  # create new disk if disk size mismatch
+                or (self._disk is None and self.disk_mask)):  # or if disk does not yet exist
 
-            x = self.disk * x
+            self._disk = make_disk(x.shape[-1]).to(x.device)
+            x = self._disk * x
 
         return x
+
+    def display_filters(self, zoom=5., **kwargs):
+        """Displays convolutional filters of FrontEnd model
+        Parameters
+        ----------
+        zoom: float
+            Magnification factor for po.imshow()
+        **kwargs:
+            Keyword args for po.imshow
+        Returns
```

Comment:
Filters can be complex-valued, right? Should we explicitly add the `plot_complex` keyword to this function?

Reply:
done.